### DISCUSSION TITLE
so-rule modify add - parsing problem

### DISCUSSION THREAD

**Discussion by tomaszniezbedny** [Security-Onion-Solutions/securityonion#5077] [Unsupported Versions]
Hi all,
I noticed a problem in rule tunning when a "$" sign appears in rules. Lets say I have following rule:
`alert http $EXTERNAL_NET any -> $HOME_NET any (msg:"ET SHELLCODE Common 0a0a0a0a Heap Spray String"; flow:established,to_client; content:"0a0a0a0a"; nocase; reference:url,www.darkreading.com/vulnerabilities---threats/heap-spraying-attackers-latest-weapon-of-choice/d/d-id/1132487; classtype:shellcode-detect; sid:2012252; rev:5; metadata:created_at 2011_02_02, former_category SHELLCODE, updated_at 2019_09_27;)`

And I want to add an exeption to it - IP address 23.197.105.116. Ultimately I want to have:
`alert http [$EXTERNAL_NET,!23.197.105.116] any -> $HOME_NET any ...`

When I run so-rule modify I get:
`so-rule modify add 2012252 "$EXTERNAL_NET any" "[$EXTERNAL_NET,!23.197.105.116] any"`
`so-rule modify add 2012252 "$EXTERNAL_NET any" "[$EXTERNAL_NET,sudo bash so-setup-network.197.105.116] any"`
`Configuration updated. Would you like to apply your changes now? (y/N) y`

`2021-08-05 13:36:29,026 - <ERROR> - Modification of rule [1:2012252] results in invalid rule: alert http $EXTERNAL_NET[,sudo bash so-setup-network.197.105.116] any -> $HOME_NET[,sudo bash so-setup-network.197.105.116] any (msg:"ET SHELLCODE Common 0a0a0a0a Heap Spray String"; flow:established,to_client; content:"0a0a0a0a"; nocase; reference:url,www.darkreading.com/vulnerabilities---threats/heap-spraying-attackers-latest-weapon-of-choice/d/d-id/1132487; classtype:shellcode-detect; sid:2012252; rev:5; metadata:created_at 2011_02_02, former_category SHELLCODE, updated_at 2019_09_27;)
2021-08-05 13:36:29,115 - <INFO> - Disabled 77 rules.`

Execution of:
`so-rule modify list`
`Modified rules + modifications:`
`  - 2012252 " any" "[,sudo bash so-setup-network.197.105.116] any"`

I tried to escape with `\$ `- not works. So I wanted to add a line in /opt/so/conf/idstools/etc/modify.conf:
`2012252 "$EXTERNAL_NET any" "[$EXTERNAL_NET,!23.197.105.116] any"`

I ran so-rule-update but it does not modify the config. I suppose $ mark is problematic. Any ideas?

**Comment by felixm223** [comment]
Wondering if you ever found a solution to this, I having the same issue even using the escape "\" in the first var. as per the training video

https://youtu.be/8TfnocUuQnM

**Comment by reyesj2** [comment]
Before the fix ensure you remove the incorrect modifications. Check ``` so-rule modify list ``` then remove the bad ones with ```so-rule modify remove <copy from so-rule modify list>```

Replace your double quotes with single quotes
```
so-rule modify add 2012252 '$EXTERNAL_NET any' '[$EXTERNAL_NET,!23.197.105.116] any'
```

**Reply by felixm223** [reply]
Thank you for the correction, I found the to edit the minions file so it that got me going.

https://docs.securityonion.net/en/2.3/managing-alerts.html#modify-the-sid
